### PR TITLE
Improve Ellipse Tool

### PR DESCRIPTION
**Pull request by pkupper**
Implement Ellipse tool as specified in #89.

---
This change is [Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/98)

### CODE REVIEW

File: core/document/src/layers/ellipse.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	fn render(&mut self, svg: &mut String) {
+		let _ = write!(
+			svg,
+			r#"<ellipse cx="0" cy="0" rx="{}" ry="{}" transform="translate({} {}) rotate({})" {} />"#,
+			self.shape.radii().x,
+			self.shape.radii().y,
+			self.shape.center().x,
+			self.shape.center().y,
+			self.shape.rotation().to_degrees(),
+			self.style.render(),
```
Comment (T0mstone) [inline core/document/src/layers/ellipse.rs:31]: This would be a lot more readable with named format arguments:
```
… rx="{rx}" ry="{ry}" transform="translate({cx} {cy}) rotate({rot})" {}/>"#, rx=self.shape.radii().x, ry=self.shape.radii().y, …
```
also the multiple calls to `radii` and `center` should be factored out. You can even do it like this:
```
let kurbo::Vec2 { x: rx, y: ry } = self.shape.radii();
let kurbo::Point { x: cx, y: cy } = self.shape.center();
```
note that format args don't have shorthand syntax afaik, so you still have to write `cx=cx` etc.

File: core/document/src/document.rs [hunk 2]
```diff
@@ -197,6 +197,24 @@ impl Document {
...
+				cx,
+				cy,
+				rx,
+				ry,
+				rot,
+				style,
+			} => {
+				self.add_layer(
+					&path,
+					Layer::new(LayerDataTypes::Ellipse(layers::Ellipse::new(kurbo::Point::new(cx, cy), kurbo::Vec2::new(rx, ry), rot, style))),
```
Comment (T0mstone) [inline core/document/src/document.rs:212]: since `layers::Ellipse::new` uses `impl Into<kurbo::{Point, Vec2}>`, this code could be made simpler by omitting those `new` calls in favor of `(f64, f64)`: 
```
layers::Ellipse::new((cx, cy), (rx, ry), …)
```
This change should then also be made in the other match arms.
If not, we should remove the `impl Into` from those `new` function's arguments.
What do you think, @Keavon?

File: core/editor/src/tools/ellipse.rs [hunk 3]
```diff
@@ -58,37 +62,112 @@ impl Fsm for EllipseToolFsmState {
...
 	}
 }
+
+fn make_operation(data: &EllipseToolData, tool_data: &DocumentToolData) -> Operation {
+	let x0 = data.drag_start.x as f64;
+	let y0 = data.drag_start.y as f64;
+	let x1 = data.drag_current.x as f64;
+	let y1 = data.drag_current.y as f64;
+
+	let (cx, cy) = if data.center_around_cursor { (x0, y0) } else { ((x0 + x1) * 0.5, (y0 + y1) * 0.5) };
```
Comment (T0mstone) [inline core/editor/src/tools/ellipse.rs:138]: Is there a reason this uses `* 0.5` instead of the conceptually simpler `/ 2.0`?

File: core/editor/src/tools/ellipse.rs [hunk 4]
```diff
@@ -58,37 +62,112 @@ impl Fsm for EllipseToolFsmState {
...
+	let (mut rx, mut ry) = if data.constrain_to_circle {
+		let r = f64::max((x1 - x0).abs(), (y1 - y0).abs());
+		(r, r)
+	} else {
+		((x1 - x0).abs(), (y1 - y0).abs())
+	};
+
+	if !data.center_around_cursor {
+		rx *= 0.5;
+		ry *= 0.5;
```
Comment (T0mstone) [inline core/editor/src/tools/ellipse.rs:149]: same as above: why not use `/= 2.0`?

File: core/editor/src/tools/ellipse.rs [hunk 5]
```diff
@@ -58,37 +62,112 @@ impl Fsm for EllipseToolFsmState {
...
+fn make_operation(data: &EllipseToolData, tool_data: &DocumentToolData) -> Operation {
+	let x0 = data.drag_start.x as f64;
+	let y0 = data.drag_start.y as f64;
+	let x1 = data.drag_current.x as f64;
+	let y1 = data.drag_current.y as f64;
+
+	let (cx, cy) = if data.center_around_cursor { (x0, y0) } else { ((x0 + x1) * 0.5, (y0 + y1) * 0.5) };
+
+	let (mut rx, mut ry) = if data.constrain_to_circle {
+		let r = f64::max((x1 - x0).abs(), (y1 - y0).abs());
```
Comment (T0mstone) [inline core/editor/src/tools/ellipse.rs:141]: maybe this should use the min instead?

File: core/editor/src/tools/ellipse.rs [hunk 6]
```diff
@@ -58,37 +62,112 @@ impl Fsm for EllipseToolFsmState {
...
+			path: vec![],
+			insert_index: -1,
+			cx,
+			cy,
+			rx,
+			ry,
+			rot: 0.0,
+			style: style::PathStyle::new(None, Some(style::Fill::new(tool_data.primary_color))),
+		}
+	}
```
Comment (T0mstone) [inline core/editor/src/tools/ellipse.rs:172]: checking `data.constrain_to_circle` twice seems kind of clunky to me.
Maybe this `if` should include the radius calculating code...

File: core/editor/src/tools/rectangle.rs [hunk 7]
```diff
@@ -72,8 +72,6 @@ impl Fsm for RectangleToolFsmState {
 				RectangleToolFsmState::LmbDown
 			}
 			(RectangleToolFsmState::LmbDown, Event::LmbUp(mouse_state)) => {
-				let r = data.drag_start.distance(&mouse_state.position);
-				log::info!("draw rectangle with radius: {:.2}", r);
```
Comment (T0mstone) [inline core/editor/src/tools/rectangle.rs:76]: why is this included in this PR? it has nothing to do with it

**Comment by T0mstone** [comment]
oh and sorry for sounding a bit harsh at times

File: core/document/src/document.rs (hunk 2 above)
Comment (Keavon) [reply core/document/src/document.rs:212]: The tuple form makes it shorter and easier to read.

File: core/editor/src/tools/ellipse.rs (hunk 3 above)
Comment (Keavon) [reply core/editor/src/tools/ellipse.rs:138]: Do they get optimized into the same assembly? I've always understood that multiplication is faster than division. To my eyes, they look equally readable. I wouldn't bother changing it.

File: core/editor/src/tools/ellipse.rs (hunk 5 above)
Comment (Keavon) [reply core/editor/src/tools/ellipse.rs:141]: Author's choice, see my comment above.

File: core/editor/src/tools/ellipse.rs (hunk 3 above)
Comment (T0mstone) [reply core/editor/src/tools/ellipse.rs:138]: Yes they do: https://godbolt.org/z/1rso37Tqv

File: core/editor/src/tools/ellipse.rs (hunk 5 above)
Comment (T0mstone) [reply core/editor/src/tools/ellipse.rs:141]: It's not equivalent though.`max` means the radius will be the larger of the two side lengths,
`min` means the shorter. I would think the shorter to be better since then the circle fits into the drawn rectangle and also since affinity does it like that iirc

File: core/editor/src/tools/ellipse.rs (hunk 3 above)
Comment (T0mstone) [reply core/editor/src/tools/ellipse.rs:138]: It is [different assermbly](https://godbolt.org/z/1rso37Tqv) but both only a single instruction and I'm not qualified to talk about the performance difference

File: core/editor/src/tools/ellipse.rs (hunk 5 above)
Comment (T0mstone) [reply core/editor/src/tools/ellipse.rs:141]: and I think you meant to reply to the previous (up by one) comment

Comment (Keavon) [reply core/editor/src/tools/ellipse.rs:141]: Sorry, somehow this comment ended up in the wrong thread. I was replying in reference to `/= 2.0;`

File: core/editor/src/tools/ellipse.rs (hunk 4 above)
Comment (Keavon) [reply core/editor/src/tools/ellipse.rs:149]: Author's choice, see my comment above.

**Review by Keavon — APPROVED** [review]

File: core/editor/src/tools/ellipse.rs (hunk 3 above)
Comment (Keavon) [reply core/editor/src/tools/ellipse.rs:138]: Food for thought: https://stackoverflow.com/questions/4125033/floating-point-division-vs-floating-point-multiplication

Comment (TrueDoctor) [reply core/editor/src/tools/ellipse.rs:138]: why not just use
```rust
#[repr(C)]
union AwesomeFloat {
    u: u32,
    f: f32,
}
let fl = AwesomeFloat { f: 3.141};
unsafe { fl.u += 1<<24};
unsafe { fl.u -= 1<<24};
```
while we're already at it?

File: core/document/src/layers/ellipse.rs (hunk 1 above)
Comment (pkupper) [reply core/document/src/layers/ellipse.rs:31]: It looks like format arguments requires the unstable feature `format_args_capture`. I'll factor out the `radii`and `center` calls but leave the arguments for now.

File: core/editor/src/tools/ellipse.rs (hunk 5 above)
Comment (pkupper) [reply core/editor/src/tools/ellipse.rs:141]: GIMP uses max for their ellipse tool. That way, the starting and ending cursor positions are both on the bounding square of the circle. This will feel more intuitive when overlays are in place. I would however be fine with both.

File: core/editor/src/tools/ellipse.rs (hunk 3 above)
Comment (TrueDoctor) [reply core/editor/src/tools/ellipse.rs]: The performance of / 2.0 is most likely worse, but the assembly does not differ when compiled with optimizations.
I'd say the slight difference in debug mode is most likely completely negligible
